File: longest-consecutive-sequence/shaun-lt-set.cpp
//lt use of set, construct the consecutive sequence +1-1 with the set membership check
//shaun read and understand
//runtime 8ms, faster than 97.60%, mem less than 57%
class Solution {
public:
    int longestConsecutive(vector<int>& nums) {
        unordered_set<int> set1;
        for (int i : nums) {
            set1.insert(i);
        }

        int longestStreak = 0;

        for (int i : set1) {
            if (!set1.count(i-1)) {
                int currentNum = i;
                int currentStreak = 1;

                while (set1.count(currentNum+1)) {
                    currentNum += 1;
                    currentStreak += 1;
                }

                longestStreak = max(longestStreak, currentStreak);
            }
        }

        return longestStreak;
    }
    
};

File: longest-consecutive-sequence/shaun-lt-set.java
/*
Runtime: 3 ms, faster than 91.51% of Java online submissions for Longest Consecutive Sequence.
Memory Usage: 40 MB, less than 13.80% of Java online submissions for Longest Consecutive Sequence.

lt set idea, check all number for possible start, stretch it
shaun understand
*/
class Solution {
    public int longestConsecutive(int[] nums) {
        Set<Integer> num_set = new HashSet<Integer>();
        for (int num : nums) {
            num_set.add(num);
        }

        int longestStreak = 0;

        //check each number if it can be a start of +1 sequence 
        for (int num : num_set) {
            //num should be start, don't want num-1 in the set
            if (!num_set.contains(num-1)) { //comment out works, but much slower
                int currentNum = num;
                int currentStreak = 1;
                
                //strech the sequence as much as possible
                while (num_set.contains(currentNum+1)) {
                    currentNum++;
                    currentStreak++;
                }
                //update longest
                longestStreak = Math.max(longestStreak, currentStreak);
            }
        }

        return longestStreak;
    }
}
